### PR TITLE
support redirects that require cookies

### PR DESCRIPTION
**Pull request by dabear**
We have a case where we're running blackbox_exporter on a linux host, and we want to monitor a Windows machine running IIS web server with the DotNetNuke (DNN) framework.

If you hit the website root path for the first time, you will get a 302 Moved Error Message together with a cookie and a redirect back to it self. If you follow that redirect with the cookie, you will get a 200 OK.

Blackbox_exporter would previously not allow cookies, and therefore DNN would always return 302. With this fix, blackbox_exporter will follow the redirect with a cookiejar and the DNN site would return a 200 OK message.

This can also be seen with curl by toggling the use of cookies or not:
```
bjorninge@bjorninge-lenovo:~/code/blackbox_exporter$ curl  https://mydnnsite.example.com/
<html><head><title>Object moved</title></head><body>
<h2>Object moved to <a href="https://mydnnsite.example.com//">here</a>.</h2>
</body></html>

bjorninge@bjorninge-lenovo:~/code/blackbox_exporter$ curl -I https://mydnnsite.example.com/
HTTP/1.1 302 Found
Content-Length: 153
Location: https://mydnnsite.example.com/
X-Redirect-Reason: Requested SplashPage Requested
Set-Cookie: dnn_IsMobile=False; path=/; HttpOnly
Set-Cookie: SplashPageView=true; path=/; HttpOnly
Date: Mon, 10 Jul 2017 13:24:18 GMT


```
By specifying cookies you will get a 200 OK response

```
bjorninge@bjorninge-lenovo:~/code/blackbox_exporter$ curl -I -L  -c cookies.txt  https://mydnnsite.example.com/
HTTP/1.1 302 Found
Content-Length: 153
Location: https://mydnnsite.example.com/
X-Redirect-Reason: Requested SplashPage Requested
Set-Cookie: dnn_IsMobile=False; path=/; HttpOnly
Set-Cookie: SplashPageView=true; path=/; HttpOnly
Date: Mon, 10 Jul 2017 13:25:30 GMT

HTTP/1.1 200 OK
Cache-Control: no-cache
Pragma: no-cache
Content-Length: 9326
Content-Type: text/html; charset=utf-8
Expires: -1
Set-Cookie: dnn_IsMobile=False; path=/; HttpOnly
Set-Cookie: .ASPXANONYMOUS=gLr6ObG8FPix2EndUblI9h2pZekfwkKzfRKTZrUoTl2p-fLgl_vMMkcCf5ApKiC0rEZ-UoUAJyZl3FLwTbbQWy-zMZ53jQ8rOJYHR7juh23BpFnA0; expires=Mon, 18-Sep-2017 00:05:30 GMT; path=/; HttpOnly
X-AspNetMvc-Version: 5.1
X-UA-Compatible: IE=edge
Set-Cookie: dnn_IsMobile=False; path=/; HttpOnly
Set-Cookie: .ASPXANONYMOUS=gLr6ObG8FPix2EndUblI9h2pZekfwkKzfRKTZrUoTl2p-fLgl_vMMkcCf5ApKiC0rEZ-UoUAJyZl3FLwTbbQWy-zMZ53jQ8rOJYHR7juh23BpFnA0; expires=Mon, 18-Sep-2017 00:05:30 GMT; path=/; HttpOnly
Set-Cookie: language=en-US; path=/; HttpOnly
Date: Mon, 10 Jul 2017 13:25:30 GMT

```

This patch fixes Blackbox exporter to make it able to monitor these kinds of web sites that require HTTP cookies.

http://pix.toile-libre.org/?img=1499692971.png

### CODE REVIEW

File: http.go [hunk 1]
```diff
@@ -120,6 +123,13 @@ func probeHTTP(target string, module Module, registry *prometheus.Registry) (suc
 		log.Errorf("Error generating HTTP client: %v", err)
 		return false
 	}
+
+	//log.Infoln("Created new empty jar")
+	jar, err := cookiejar.New(&cookiejar.Options{PublicSuffixList: publicsuffix.List})
+	if err != nil {
+		log.Fatal(err)
```
Comment (brian-brazil) [inline http.go]: Don't kill the entire process if this fails, just log and return false

File: http.go [hunk 2]
```diff
@@ -120,6 +123,14 @@ func probeHTTP(target string, module Module, registry *prometheus.Registry) (suc
 		log.Errorf("Error generating HTTP client: %v", err)
 		return false
 	}
+
+	//log.Infoln("Created new empty jar")
+	jar, err := cookiejar.New(&cookiejar.Options{PublicSuffixList: publicsuffix.List})
+	if err != nil {
+		log.Error(err)
```
Comment (brian-brazil) [inline http.go]: Can you use Errorf and add a bit of context?

File: http.go [hunk 3]
```diff
@@ -120,6 +123,14 @@ func probeHTTP(target string, module Module, registry *prometheus.Registry) (suc
 		log.Errorf("Error generating HTTP client: %v", err)
 		return false
 	}
+
+	//log.Infoln("Created new empty jar")
```
Comment (brian-brazil) [inline http.go]: Please remove the debug line.

**Comment by dabear** [comment]
I think unit testing for this would be difficult, because it will require access to a public DNN site.
How would I go about adding the vendoring? Any docs? I am unfamiliar how this works on the Go world

**Comment by brian-brazil** [comment]
For vendoring we use govendor. `govendor add +external` should do it.

For testing, I'd create a test http server that has a redirecting endpoint that returns a cookie and then verifies that cookie is set on the endpoint that's redirected to. There's existing http tests using a test http server.

**Comment by dabear** [comment]
I've started implementing a cookie test, but cannot figure out how to get access to the response cookies?
Would you consider merging this even if I can't figure out the unit test?
```
func TestCookieSet(t *testing.T) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		expiration := time.Now().Add(365 * 24 * time.Hour)
		cookie := http.Cookie{Name: "somecookie", Value: "cookie", Expires: expiration}
		http.SetCookie(w, &cookie)
		http.Redirect(w, r, "/noredirect", http.StatusFound)
	}))
	defer ts.Close()

	// Follow redirect, should succeed with 200.
	recorder := httptest.NewRecorder()
	registry := prometheus.NewRegistry()

	result := probeHTTP(ts.URL,
		Module{Timeout: time.Second, HTTP: HTTPProbe{}}, registry)
	/*
		somehow get access to probeHTTP's client and check resp.Cookies() here how??
	*/
	//t.Fatalf("test cookie fail ex")
	body := recorder.Body.String()
	if !result {
		t.Fatalf("Redirect test failed unexpectedly, got %s", body)
	}

}
```

**Comment by brian-brazil** [comment]
> I've started implementing a cookie test, but cannot figure out how to get access to the response cookies?

It's `response.Cookies()`, and similarly for requests.

> Would you consider merging this even if I can't figure out the unit test?

I would prefer not to have untested features in the codebase.

**Comment by dabear** [comment]
I'd love some more help on that, because I can't seem to access any variable "response from http_test.go! My go skills are not that great, so I need to have some more info on how to do this..

./http_test.go:79: undefined: response in response.Cookies
alt text

**Comment by SuperQ** [comment]
Yes, we can help a bit with the testing and other stuff.

I'm going to open a separate PR to add the vendoring you need here.  That will make this easier to review/merge.

**Comment by dabear** [comment]
Thanks, but I don't see how publicsuffix vendoring will help me get access to the response cookies? :)

**Comment by SuperQ** [comment]
@dabear There were a couple issues with the vendoring you added.  You should be able to rebase your fork with upstream now.

**Comment by dabear** [comment]
is it correct now or not? Having a hard time wrapping my head around what I should do...

**Comment by SuperQ** [comment]
It looks like you've been adding commits to your fork's master.  Generally this just leads to git history pain.

I suggest you rebase your fork cleanly against upstream and make your changes to a branch.  This makes things like `git rebase` work much more smoothly.

**Comment by dabear** [comment]
Closing this, I am no longer able to create requests with cookies correctly. Further investigations seems to point to a change in blackbox_exporter that sets targeturl.Host to an IP. This leads to errors on our 
endpoints:
 2017-09-15 14:16:10 10.25.49.5 GET / - 443 - 86.62.191.174 Go-http-client/1.1 http://[195.225.7.100]/ 500 0 0 7 -

**Pull request closed without merging**